_posts/2018-05-29-mono-on-aix-and-ibm-i.md

```diff
@@ -0,0 +1,273 @@
+yet manageable codebase due to a low number of dependencies, use of plain C99,
+and an emphasis on portability. Most common processor architectures in use are
+supported by the code generator, though more obscure ISAs will have some caveats.
+
+Pretty much all of the work you do will be twiddling #ifdefs to accommodate for
+the target platform’s quirks; such as missing or different preprocessor
+definitions and functions, adding the platform to definitions so it is
+supported by core functionality, and occasionally having to tweak the runtime
+or build system to handle when the system does something completely differently
+than others. In the case of AIX and IBM i, I had to do all of these things.
+
+## [Where would I be without IBM?](https://www.youtube.com/watch?v=dYBfdtMAD2Q)
+
+For some background on what needed to happen, we can start by giving some
+background on our target platforms.
+
+Both of our targets run on 64-bit PowerPC processors in big endian mode.
+[Mono does support PowerPC](/docs/about-mono/supported-platforms/powerpc),
+and Bernhard Urban maintains it. What is odd about the calling conventions on
+AIX (shared occasionally by Linux) is the use of *function descriptors*, which
+means that pointers to functions do not point to code, but instead point to
+metadata about them. This can cause bugs in the JIT if you are not careful to
+consume or produce function descriptors instead of raw pointers when needed.
+Because the runtime is better tested on 64-bit PowerPC, and machines are fast
+enough that the extra overhead is not significant, we always build a 64-bit
+runtime.
+
+In addition to a strange calling convention, AIX also has a different binary
+format - that means that currently, the ahead-of-time compiler does not work.
+While most Unix-like operating systems use ELF, AIX (and by extension, IBM i
+for the purposes of this port) use XCOFF, a subset of the Windows PE binary
+format.
+
+AIX is a Unix (descended from the System V rather than the BSD side of the
+family) that runs on PowerPC systems. Despite being a Unix, it has some quirks
+of its own, that I will describe in this article.
+
+### Unix? What’s a Unix?
+
+IBM i (formerly known as i5/OS or OS/400) is decidedly not a Unix. Unlike Unix,
+it has an object-based filesystem where all objects are mapped into a single
+humongous address space, backed on disk known as single level storage – real
+main storage (RAM) holds pages of objects "in use" and acts as a cache for
+objects that reside permanently on disk. Instead of flat files, IBM i uses
+database tables as the means to store data. (On IBM i, all files are database
+tables, and a file is just one of the “object types” supported by IBM i; others
+include libraries and programs.) Programs on IBM i are not simple native
+binaries, but instead are "encapsulated" objects that contain an intermediate
+form, called Machine Interface instructions, (similar to MSIL/CIL) that is then
```

Comment:
We should output this instruction set instead?

Reply:
That'd require a port to the ILE; the native i environment, which is not remotely POSIX and runs as MI itself, with no knowledge of the host CPU. This current port is targets and runs on normal AIX for ppc64be, but that's only because we're really targeting PASE which is compatible with that subset.

I've considered it, but it is a lot of work to adapt to the native environment and would be AOT only, because there is no JIT of MI. FWIW, MI has to be compiled through the `QPRCRTPG` API.

Reply:
I'll consider adding a paragraph about a theoretical port to the ILE, FWIW.